### PR TITLE
Enable tests in file_system

### PR DESCRIPTION
**Pull request by keepsimple1**
- Started to update doc for this crate.
- Moved a doc test in a real test case.
- Found one test file `file_system` was not enabled.  Enabled it.
- Fixed a couple bugs after the above test was enabled.

### CODE REVIEW

File: radicle-surf/t/src/file_system.rs [hunk 1]
```diff
@@ -50,6 +49,49 @@ mod directory {
...
+        assert!(root_contents[6].is_directory());
+        assert_eq!(root_contents[3].name(), "bin");
+        assert_eq!(root_contents[4].name(), "src");
+        assert_eq!(root_contents[5].name(), "text");
+        assert_eq!(root_contents[6].name(), "this");
+
+        let src = root
+            .find_directory(Path::new("src"), &repo)
+            .unwrap()
+            .unwrap();
```
Comment (keepsimple1) [inline radicle-surf/t/src/file_system.rs:87]: I didn't like the double unwrap. I think it is probably a better interface if `find_directory` returns `Result<>`, not `Result<Option<>>`. 

If this directory was not found, we can return an error like `NotFound`.

File: radicle-surf/examples/browsing.rs [hunk 2]
```diff
@@ -51,10 +51,10 @@ fn main() {
 fn print_directory(d: &Directory, repo: &Repository, indent_level: usize) {
     let indent = " ".repeat(indent_level * 4);
     println!("{}{}/", &indent, d.name());
-    for entry in d.entries(repo).unwrap().entries() {
+    for entry in d.entries(repo).unwrap() {
```
Comment (keepsimple1) [inline radicle-surf/examples/browsing.rs:54]: Double `entries()` in the existing code seems confusing, so I added `impl Iterator` for  `Entries` to support this simpler usage. 

Another option is to rename the first `entries()` method, maybe.

File: radicle-surf/src/file_system/directory.rs [hunk 3]
```diff
@@ -327,31 +350,31 @@ impl Directory {
...
+            .get_path(path)
             .map(Some)
             .or_matches::<git2::Error, _, _>(is_not_found_err, || Ok(None))?;
+        let parent = path
+            .parent()
+            .ok_or_else(|| git::Error::PathNotFound(path.to_path_buf()))?;
 
         Ok(entry
-            .and_then(|entry| Entry::from_entry(&entry, path.as_ref().to_path_buf()).transpose())
+            .and_then(|entry| Entry::from_entry(&entry, parent.to_path_buf()).transpose())
```
Comment (keepsimple1) [inline radicle-surf/src/file_system/directory.rs:371]: This was what apparently failed the tests.

File: radicle-surf/src/file_system/directory.rs [hunk 4]
```diff
@@ -327,31 +350,31 @@ impl Directory {
     /// Find the [`Entry`] found at `path`, if it exists.
     pub fn find_entry<P>(
         &self,
-        path: &P,
+        path: P,
```
Comment (FintanH) [inline radicle-surf/src/file_system/directory.rs:353]: Why did you change taking by reference here?

File: radicle-surf/t/src/file_system.rs (hunk 1 above)
Comment (FintanH) [reply radicle-surf/t/src/file_system.rs:87]: I'm on the fence about having a `NotFound` error for a `find_*` function. I generally prefer that the caller has the option to handle that case explicitly through the `Option`, or they use their own `NotFound` error and `ok_or`. Another option is to have a `try_find_file`, but that still means we end up hiding a `NotFound` in our error variants.

The fact that this is a double unwrap is only because we're in a testing environment.

**Review by FintanH — DISMISSED** [review]
Nice, thanks for catching the bug! Left some comments below :)

File: radicle-surf/t/src/file_system.rs (hunk 1 above)
Comment (keepsimple1) [reply radicle-surf/t/src/file_system.rs:87]: I will leave it as is for now.

File: radicle-surf/src/file_system/directory.rs (hunk 4 above)
Comment (keepsimple1) [reply radicle-surf/src/file_system/directory.rs:353]: Because `P` is already a reference: `P: AsRef<Path>`.  And in practice `Path::new()` returns `&Path`.

Comment (keepsimple1) [reply radicle-surf/src/file_system/directory.rs:353]: Because `P` is already a reference: `P: AsRef<Path>`.  And in practice `Path::new()` returns `&Path`.

Comment (keepsimple1) [reply radicle-surf/src/file_system/directory.rs:387]: It is also consistent with the existing `find_directory`

Comment (FintanH) [reply radicle-surf/src/file_system/directory.rs:387]: > Because `P` is already a reference: `P: AsRef<Path>`. And in practice `Path::new()` returns `&Path`.

That's not necessarily what `P: AsRef<Path>` means. If `P: AsRef<Path>` that implies the following code:
```rust
 // for some P
impl AsRef<Path> for P {
  fn as_ref(&self) -> &Path {
    /* code */
  }
}
```

The `&self` means that we have a reference of type `P`. So when we take ownership of `P` in the way you're doing above that satisfies the borrow checker since owned implies borrowed. That's quite restrictive. Say if I wanted to use the same `path` for multiple functions including `find_entry`, I would have to `clone` to do so -- when the `clone` is not really needed.

I must have missed this with `find_directory` so I would say we change `find_entry` back and fix `find_directory` :)

Comment (keepsimple1) [reply radicle-surf/src/file_system/directory.rs:387]: >  Say if I wanted to use the same path for multiple functions including find_entry, I would have to clone to do so -- when the clone is not really needed.

In practice, `Path::new()` returns `&Path`, and you probably always have `&Path` not `Path` and no need to clone `Path`.

Comment (FintanH) [reply radicle-surf/src/file_system/directory.rs]: > > Say if I wanted to use the same path for multiple functions including find_entry, I would have to clone to do so -- when the clone is not really needed.
> 
> In practice, [`Path::new()`](https://doc.rust-lang.org/std/path/struct.Path.html#method.new) returns `&Path`, and you probably always have `&Path` not `Path` and no need to clone `Path`.
> 
> I will make the change.

Right, so even more reason to have a borrowed parameter :smile: You would have to use `to_path_buf` to make it owned :)

**Review by FintanH — APPROVED** [review]
